### PR TITLE
CB-15260: DES not deleted on Azure cloud if encryption key is wrong

### PR DESCRIPTION
**Pull request by juhi-09**
CB-15260: DES not deleted on Azure cloud if encryption key is wrong

This PR fixes following issues- 
1. Checks the existence of encryptionKeyUrl to delete the encryption resources. 
2. Saves azureEncryptionParameters correctly while adding the encryption resources for existing environments. "environmentService.save(environment)" was not able to save the encryptionKeyUrl as it is a secret. As a r
esult, saving it using AzureParameters entity.
3. Checks for the existence of keyVault before creating the disk encryption set. A check has been added before removing the disk encryption set's access policy from keyVault - to handle the case where keyVault is deleted/access changed but environment still exists. Without this existence check it would through an NPE.

Tests -
1. Updated the existing test.
2. Tested the updation of encryptionKeyUrl for existing environment with both wrong key and correct key.

```
juhig@MacBook-Pro thunderhead % ./clients/cdpcli/cdp.sh environments update-azure-encryption-resources \
    --environment testcmk \
    --encryption-key-url 'https://juhig.vault.azure.net/keys/juhig-key/c648437323ad45e59041fa1d5e9307d7' \
    --encryption-key-resource-group-name 'juhig'
2021-12-03 16:24:45,234 - MainThread - cdpcli.clidriver - WARNING - You are running an INTERNAL release of the CDP CLI, which has different capabilities from the standard public release. Find the public release at: https://pypi.org/project/cdpcli/

An error occurred: {&quot;message&quot;:&quot;vaultName \&quot;juhig\&quot; either does not exists or user does not have permissions to access it. Kindly check if the vault &amp; encryption key exists and correct encryption key url is specified.&quot;,&quot;payload&quot;:null} (Status Code: 400; Error Code: INVALID_ARGUMENT; Service: environments; Operation: updateAzureEncryptionResources; Request ID: 426a7b76-de9d-49f1-84f7-c9b5e76a47e3;)```

### CODE REVIEW

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/deletion/handler/ResourceEncryptionDeleteHandler.java [hunk 1]
```diff
@@ -60,12 +58,7 @@ public void accept(Event<EnvironmentDeletionDto> environmentDtoEvent) {
         try {
             environmentService.findEnvironmentById(environmentDto.getId()).ifPresent(environment -> {
                 if (AZURE.name().equalsIgnoreCase(environmentDto.getCloudPlatform())) {
-                    String diskEncryptionSetId = Optional.ofNullable(environmentDto.getParameters())
-                            .map(ParametersDto::getAzureParametersDto)
-                            .map(AzureParametersDto::getAzureResourceEncryptionParametersDto)
-                            .map(AzureResourceEncryptionParametersDto::getDiskEncryptionSetId)
-                            .orElse(null);
-                    if (StringUtils.isNotEmpty(diskEncryptionSetId) &&
+                    if (entitlementService.isAzureDiskSSEWithCMKEnabled(environmentDto.getAccountId()) &&
```
Comment (lajosrodek) [inline environment/src/main/java/com/sequenceiq/environment/environment/flow/deletion/handler/ResourceEncryptionDeleteHandler.java:61]: As noted in the Jira ticket, I think the right approach is to check for the presence of `AzureResourceEncryptionParametersDto::getEncryptionKeyUrl` instead of `AzureResourceEncryptionParametersDto::getDiskEncryptionSetId`. That logic used to be present since #10740, but it got later reverted in #11225 for an unknown reason.

While the entitlement check may seem as a quick-and-dirty fix, it is quite inefficient. With this, deletion will be always attempted if the entitlement is granted, even if the DTO does not specify any encryption key URL at all.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Please see my comments in the parent Jira ticket. Besides the condition below, I suggest fixing the NPE in `AzureClient.grantKeyVaultAccessPolicyToServicePrincipal()` as well, either here or in a separate PR.

**Comment by lajosrodek** [comment]
Upon a closer look, `AzureClient` has two potential NPE scenarios:
* [grantKeyVaultAccessPolicyToServicePrincipal()](https://github.com/hortonworks/cloudbreak/blob/38b7f208d45d6a5a0c649ec8888ba1724de99fe9/cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java#L990-L991) (mentioned above)
* [removeKeyVaultAccessPolicyFromServicePrincipal()](https://github.com/hortonworks/cloudbreak/blob/38b7f208d45d6a5a0c649ec8888ba1724de99fe9/cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java#L1003-L1004)

In both cases, the fragile logic looks like
```
                    .getByResourceGroup(resourceGroupName, vaultName)
                    .update()
```
Here, `getByResourceGroup()` is invoked for the result of `azure.vaults()` (which is known to be non-`null`). On the other hand, based on the error mentioned in the parent Jira ticket, I suspect that `getByResourceGroup()` will simply return `null` if the Key Vault does not exist, thus raising an NPE upon the `update()` invocation. This should be verified and then fixed. The fix could employ an `Optional.ofNullable().ifPresent()` construct or an additional `if ()` statement, whichever is preferred.

File: environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java [hunk 2]
```diff
@@ -183,15 +188,15 @@ private EnvironmentDto updateAzureResourceEncryptionParameters(String accountId,
...
             } catch (Exception e) {
                 throw new BadRequestException(e);
             }
             LOGGER.debug("Successfully created the Disk encryption set for the environment {}.", environmentName);
         } else {
             throw new BadRequestException(validateKey.getFormattedErrors());
         }
-        Environment saved = environmentService.save(environment);
-        return environmentDtoConverter.environmentToDto(saved);
+        return environmentDtoConverter.environmentToDto(environment);
```
Comment (lajosrodek) [inline environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java:199]: Why this change?

File: environment/src/main/java/com/sequenceiq/environment/environment/repository/AzureDiskEncryptionParametersRepository.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import com.sequenceiq.cloudbreak.workspace.repository.EntityType;
+import com.sequenceiq.environment.parameters.dao.domain.AzureParameters;
+
+@EntityType(entityClass = AzureParameters.class)
+@Transactional(Transactional.TxType.REQUIRED)
+public interface AzureDiskEncryptionParametersRepository extends JpaRepository<AzureParameters, Long> {
+}
```
Comment (lajosrodek) [inline environment/src/main/java/com/sequenceiq/environment/environment/repository/AzureDiskEncryptionParametersRepository.java:13]: What is the reason for introducing this new repository? The usual approach is to treat the `Environment` entity as the single point of entry; nested child entities are not manipulated individually. Furthermore, the class name is also confusing; `AzureParameters` contains other things not related to disk encryption.

**Review by lajosrodek — CHANGES_REQUESTED** [review]

File: environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java [hunk 4]
```diff
@@ -183,15 +188,15 @@ private EnvironmentDto updateAzureResourceEncryptionParameters(String accountId,
                 CreatedDiskEncryptionSet createdDiskEncryptionSet = environmentEncryptionService.createEncryptionResources(
                         environmentDtoConverter.environmentToDto(environment));
                 azureParameters.setDiskEncryptionSetId(createdDiskEncryptionSet.getDiskEncryptionSetId());
+                azureDiskEncryptionParametersRepository.save(azureParameters);
```
Comment (lajosrodek) [inline environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java:191]: This invocation - and the whole `AzureDiskEncryptionParametersRepository` actually - is not needed; the `environmentService.save(environment)` is expected to persist the changes to `azureParameters.diskEncryptionSetId` as well.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
As I see, the NPE issues around `AzureClient` are not yet addressed here, so I assume they will be tackled in a follow-up PR.

File: environment/src/main/java/com/sequenceiq/environment/environment/repository/AzureDiskEncryptionParametersRepository.java (hunk 3 above)
Comment (bergerdenes) [reply environment/src/main/java/com/sequenceiq/environment/environment/repository/AzureDiskEncryptionParametersRepository.java:13]: we already have AzureParametersRepository anyway

File: environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java (hunk 2 above)
Comment (bergerdenes) [reply environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java:199]: ❗  why is the environment save removed?

**Review by bergerdenes — CHANGES_REQUESTED** [review]
rework please

File: environment/src/main/java/com/sequenceiq/environment/environment/repository/AzureDiskEncryptionParametersRepository.java (hunk 3 above)
Comment (juhi-09) [reply environment/src/main/java/com/sequenceiq/environment/environment/repository/AzureDiskEncryptionParametersRepository.java]: I have removed new repository, using AzureParametersRepository now. It is required because environmentService.save(environment) is not saving the encryptionKeyUrl as it is a secret. In order to set azureParameters.encryptionKeyUrl, AzureParametersRepository is used.

File: environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java (hunk 4 above)
Comment (juhi-09) [reply environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java]: `environmentService.save(environment)` stores the diskEncryptionSetId correctly. However, it does not persist the encryptionKeyUrl (which is a secret) - and is used during deletion of encryption resources. Hence, saving it using `AzureParametersRepository` now.

**Comment by juhi-09** [comment]
> As I see, the NPE issues around `AzureClient` are not yet addressed here, so I assume they will be tackled in a follow-up PR.

Added the code to handle the NPE.

File: environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java (hunk 2 above)
Comment (juhi-09) [reply environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentModificationService.java:199]: AzureParametersRepository.save(azureParameters) saves the parameters and cascades the changes to environment. So, I removed this extra save of environment.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java [hunk 5]
```diff
@@ -110,6 +110,13 @@ public CreatedDiskEncryptionSet createDiskEncryptionSet(DiskEncryptionSetCreatio
             String sourceVaultId = String.format("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.KeyVault/vaults/%s",
                     azureClient.getCurrentSubscription().subscriptionId(), vaultResourceGroupName, vaultName);
 
+            // Check for the existence of keyVault user has specified before creating disk encryption set(DES).
+            // If keyVault is wrong or user lacks permissions to access it, granting access permissions to this keyVault for DES would fail.
+            if (!azureClient.checkKeyVaultExistence(vaultResourceGroupName, vaultName)) {
+                throw new IllegalArgumentException(String.format("vaultName \"%s\" either does not exists or user does not have permissions to access it. " +
+                                "Kindly check if the vault & encryption key exists and correct encryption key url is specified.",
```
Comment (lajosrodek) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java:117]: Nits:
* `vaultName` -> `Vault with name` (not sure the customer would understood `vaultName` as it is an internal identifier extracted from the key URL)
* `url` -> `URL`

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java [hunk 6]
```diff
@@ -311,18 +318,26 @@ private void removeKeyVaultAccessPolicyFromDiskEncryptionSetServicePrincipal(Azu
...
-                return true;
-            } catch (Exception e) {
-                throw azureUtils.convertToActionFailedExceptionCausedByCloudConnectorException(e, "Removing " + description);
-            }
-        });
+
+        // Check for the existence of keyVault user has specified before removing disk encryption set's(DES) access permissions from this keyVault.
+        if (!azureClient.checkKeyVaultExistence(vaultResourceGroupName, vaultName)) {
+            LOGGER.info(String.format("vaultName \"%s\" either does not exists/have been deleted or user does not have permissions to access it. " +
+                            "Kindly check if the vault & encryption key exists and correct encryption key url is specified.",
```
Comment (lajosrodek) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java:325]: Nits:
* Possibly use `warn()` log level.
* `vaultName` -> `Vault with name` (though this is not critical, customers will not see this log)
* `url` -> `URL`
* I wonder if the "Kindly check ..." part actually makes sense. The deletion flow will still proceed in this case without attempting the key vault policy updates, and no one will notice this unless explicitly looking at logs. :smile:

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java [hunk 7]
```diff
@@ -110,6 +110,13 @@ public CreatedDiskEncryptionSet createDiskEncryptionSet(DiskEncryptionSetCreatio
             String sourceVaultId = String.format("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.KeyVault/vaults/%s",
                     azureClient.getCurrentSubscription().subscriptionId(), vaultResourceGroupName, vaultName);
 
+            // Check for the existence of keyVault user has specified before creating disk encryption set(DES).
```
Comment (lajosrodek) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java:113]: Nit: insert space before `(DES)`.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java [hunk 8]
```diff
@@ -311,18 +318,26 @@ private void removeKeyVaultAccessPolicyFromDiskEncryptionSetServicePrincipal(Azu
...
-                LOGGER.info("Removing {}.", description);
-                azureClient.removeKeyVaultAccessPolicyFromServicePrincipal(vaultResourceGroupName, vaultName, desPrincipalObjectId);
-                LOGGER.info("Removed {}.", description);
-                return true;
-            } catch (Exception e) {
-                throw azureUtils.convertToActionFailedExceptionCausedByCloudConnectorException(e, "Removing " + description);
-            }
-        });
+
+        // Check for the existence of keyVault user has specified before removing disk encryption set's(DES) access permissions from this keyVault.
```
Comment (lajosrodek) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java:322]: Nit: insert space before `(DES)`.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java [hunk 9]
```diff
@@ -984,6 +985,12 @@ public DiskEncryptionSetInner createDiskEncryptionSet(String diskEncryptionSetNa
         });
     }
 
+    public boolean checkKeyVaultExistence(String resourceGroupName, String vaultName) {
+        Vault keyVault = azure.vaults()
+                .getByResourceGroup(resourceGroupName, vaultName);
+        return (keyVault != null) ? true : false;
```
Comment (lajosrodek) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java:991]: Simplify as `return keyVault != null;`

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java [hunk 10]
```diff
@@ -984,6 +985,12 @@ public DiskEncryptionSetInner createDiskEncryptionSet(String diskEncryptionSetNa
         });
     }
 
+    public boolean checkKeyVaultExistence(String resourceGroupName, String vaultName) {
```
Comment (lajosrodek) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java:988]: I see there are some other `checkXXX()` methods as well already, but `keyVaultExists()` would sound more natural, and imply what the return value expresses.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java (hunk 9 above)
Comment (lajosrodek) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java:991]: Please wrap the method body in `return handleAuthException(() -> {...})` like done for `createDiskEncryptionSet()`. This is to ensure that the case when the Azure credential app has expired gets handled correctly.

**Review by lajosrodek — CHANGES_REQUESTED** [review]

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java [hunk 11]
```diff
@@ -984,6 +985,14 @@ public DiskEncryptionSetInner createDiskEncryptionSet(String diskEncryptionSetNa
         });
     }
 
+    public boolean keyVaultExists(String resourceGroupName, String vaultName) {
+        return handleAuthException(() -> {
+            Vault keyVault = azure.vaults()
+                    .getByResourceGroup(resourceGroupName, vaultName);
+        return keyVault != null;
```
Comment (lajosrodek) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClient.java:992]: Wrong indentation.

**Review by lajosrodek — APPROVED** [review]
Thanks for the fixes. Only one formatting note, but approving it in advance.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java [hunk 12]
```diff
@@ -110,6 +110,13 @@ public CreatedDiskEncryptionSet createDiskEncryptionSet(DiskEncryptionSetCreatio
             String sourceVaultId = String.format("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.KeyVault/vaults/%s",
                     azureClient.getCurrentSubscription().subscriptionId(), vaultResourceGroupName, vaultName);
 
+            // Check for the existence of keyVault user has specified before creating disk encryption set (DES).
+            // If keyVault is wrong or user lacks permissions to access it, granting access permissions to this keyVault for DES would fail.
+            if (!azureClient.keyVaultExists(vaultResourceGroupName, vaultName)) {
+                throw new IllegalArgumentException(String.format("Vault with name \"%s\" either does not exists or user does not have permissions to " +
```
Comment (bergerdenes) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java:116]: grammar: does not exist

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java [hunk 13]
```diff
@@ -311,18 +318,25 @@ private void removeKeyVaultAccessPolicyFromDiskEncryptionSetServicePrincipal(Azu
...
-                LOGGER.info("Removed {}.", description);
-                return true;
-            } catch (Exception e) {
-                throw azureUtils.convertToActionFailedExceptionCausedByCloudConnectorException(e, "Removing " + description);
-            }
-        });
+
+        // Check for the existence of keyVault user has specified before removing disk encryption set's (DES) access permissions from this keyVault.
+        if (!azureClient.keyVaultExists(vaultResourceGroupName, vaultName)) {
+            LOGGER.warn(String.format("Vault with name \"%s\" either does not exists/have been deleted or user does not have permissions to access it.",
```
Comment (bergerdenes) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureEncryptionResources.java:324]: grammar: does not exist

**Review by bergerdenes — APPROVED** [review]
minor grammar nits
